### DISCUSSION TITLE
Best pattern for making calls from Rust to a large Python package?

### DISCUSSION THREAD

**Discussion by l-monninger** [PyO3/pyo3#2402] [Questions] [answered]
I have a relatively large `Python` library that I'd like to call from `Rust`. 

From what I've read so far, it seems there might be two strategies:
1. Perform the [multiple compilation](https://pyo3.rs/latest/python_from_rust.html#include-multiple-python-files) myself.
2. Install the the package in my virtual environment and run it there. 

I would prefer not to have to write the multiple compilation logic myself, but [this](https://pyo3.rs/latest/python_from_rust.html#include-multiple-python-files) makes me think I might have to. 

Is there I'm missing? How should I do this?

**Comment by davidhewitt** [comment, marked as answer]
What's the entry point of your program? Do you want to ship a single executable, or are you fine to create a Python environment?

If you want a single executable, PyOxidizer has worked pretty well for many users.

If you are fine to create a Python environment, then I'd compile your Rust as a Python extension and use it as part of normal Python scripting. 

Usually the best way to call into Python would be to import it via normal mechanisms:

```rust
#[pyfunction]
fn call_some_foo(py: Python<'_>) -> PyResult<()> {
    let foo = PyModule::import(py, "foo")?;
    // call a function in the foo package
    foo.getattr("some_function")?.call0()?;
    Ok(())
}
```

**Reply by l-monninger** [reply]
Thanks for this!
